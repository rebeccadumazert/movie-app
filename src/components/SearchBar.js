import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';

import { searchMovie } from './../services/moviesDb';
import ResultSearchMovie from './ResultSearchMovie';
import ResultSearchPerson from './ResultSearchPerson';
import ResultSearchTv from './ResultSearchTv';
import './style/searchBar.css';

const MEDIA_URL = {
  person: 'credit',
  movie: 'movie',
  tv: 'tv',
};

class SearchBar extends Component {
  state = {
    userSearch: '',
    resultsSearch: [],
    highlightIndex: -1,
  };

  async fetchSearchMovie() {
    const {
      data: { results: resultsSearch },
    } = await searchMovie(this.state.userSearch);
    this.setState({
      resultsSearch,
    });
  }

  onChangeSearch = e => {
    this.setState({
      userSearch: e.target.value,
      highlightIndex: -1,
    });
  };
  clickBtnSearch = () => {
    this.fetchSearchMovie();
  };

  selectComponent = (result, index) => {
    const isHighLighted = index === this.state.highlightIndex;
    const medias = {
      movie: (
        <ResultSearchMovie
          key={result.id}
          result={result}
          isHighLighted={isHighLighted}
        />
      ),
      person: (
        <ResultSearchPerson
          key={result.id}
          result={result}
          isHighLighted={isHighLighted}
        />
      ),
      tv: (
        <ResultSearchTv
          key={result.id}
          result={result}
          isHighLighted={isHighLighted}
        />
      ),
    };
    return medias[result.media_type];
  };

  enterResearch = ({ keyCode }) => {
    if (keyCode === 13) {
      if (this.state.highlightIndex >= 0) {
        const media = this.state.resultsSearch[this.state.highlightIndex];
        this.props.history.push(`/${MEDIA_URL[media.media_type]}/${media.id}`);
      } else {
        this.fetchSearchMovie();
      }
    }
    if (keyCode === 40) {
      this.setState(({ highlightIndex, resultsSearch }) => {
        return {
          highlightIndex: Math.min(
            highlightIndex + 1,
            resultsSearch.length - 1
          ),
        };
      });
    }
    if (keyCode === 38) {
      this.setState(({ highlightIndex }) => {
        return {
          highlightIndex: Math.max(0, highlightIndex - 1),
        };
      });
    }
    if (keyCode === 27) {
      this.setState({
        resultsSearch: [],
        userSearch: '',
      });
    }
  };

  redirection = () => {
    console.log('coucou');
  };
  render() {
    return (
      <div className="searchBarContainer">
        <input
          className="inputSearch"
          value={this.state.userSearch}
          onChange={this.onChangeSearch}
          type="text"
          onKeyDown={this.enterResearch}
        />

        <button className="buttonStyle" onClick={this.clickBtnSearch}>
          SEARCH
        </button>
        <div className="selectedComponent">
          {this.state.resultsSearch.map((result, i) =>
            this.selectComponent(result, i)
          )}
        </div>
      </div>
    );
  }
}
export default withRouter(SearchBar);
