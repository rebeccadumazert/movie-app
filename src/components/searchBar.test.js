import React from 'react';
import axios from 'axios';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import SearchBar from './SearchBar';

configure({ adapter: new Adapter() });

afterEach(() => {
  // cleaning up the mess left behind the previous test
  axios.reset();
});

test('Actor renders correctly', () => {
  const tree = renderer.create(<SearchBar.WrappedComponent />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('simulate click events', () => {
  const searchBar = shallow(<SearchBar.WrappedComponent />);
  console.log(searchBar.text());
  searchBar.find('button').simulate('click');
  expect(axios.get).toHaveBeenCalledTimes(1);
});
