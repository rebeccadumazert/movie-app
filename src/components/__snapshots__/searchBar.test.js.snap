// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actor renders correctly 1`] = `
<div
  className="searchBarContainer"
>
  <input
    className="inputSearch"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
    value=""
  />
  <button
    className="buttonStyle"
    onClick={[Function]}
  >
    <img
      alt="magnifying glass"
      src="https://img.icons8.com/ios-glyphs/30/000000/search.png"
    />
  </button>
  <div
    className="selectedComponent"
  />
</div>
`;
